docs: choreography saga quickstart (#1884)
lightbend/kalix-jvm-sdk: samples/java-spring-choreography-saga-quickstart/README.md

# Choreography Saga Quickstart

This quickstart sample demonstrates how to implement a Choreography Saga in Kalix.

This project explores the usage of Event Sourced Entities, Value Entities and Actions.  
To understand more about these components, see [Developing services](https://docs.kalix.io/services/) and check 
Java-SDK [official documentation](https://docs.kalix.io/java/index.html). 

## What is a Choreography Saga?

A **Choreography Saga** is a distributed transaction pattern that helps you manage transactions across multiple services.

In the context of event-driven applications, a **Choreography Saga** is implemented as a sequence of transactions,
each of which publishes an event or state change notification that triggers the next operation in the saga.
If an operation fails because it violates a business rule, then the saga can execute a series of compensating transactions
that undo the changes made by the previous operations.

In Kalix, in addition to events from Event Sourced Entities, you can also subscribe to state changes from Value Entities.
To subscribe to events or state changes, we can use Kalix Actions with the appropriate subscription annotations.

> You can create a Choreography Saga to manage transactions across multiple entities in a single service, or across multiple services.

Author: Yes, correct. 
But I think that the quickstart should reiterate the patterns used in the quickstart and point to the documentation for more in-depth information about the pattern. 

Also, I should add links to EventSourcedEntity, ValueEntity, Action as Controller and Actions for Subscriptions. Those are the building blocks used here and we should link it back to the pages in the docs.

> You can create a Choreography Saga to manage transactions across multiple entities in a single service, or across multiple services.

Author: Not sure if what I just wrote was clear. So rephrasing it. 

I agree that we should have a section about Sagas and event about set-base consistency validation in the main documentation. And that should be more in-depth than what we mention here. 

But once we touch one or other topic in a quickstart, we give a short description about it and then we point back to the main documentation for more details.

WDYT?

> You can create a Choreography Saga to manage transactions across multiple entities in a single service, or across multiple services.

Author: Created a ticket for it.

This example implements a choreography that manages transactions across two entities in the same service.

## Cross Entity Field Uniqueness

> A common challenge in event-sourced applications is called the _Set-Based Consistency Validation_ problem. It arises when we need to ensure that a particular field is unique across all entities in the system. For example, a user may have a unique identifier (e.g. social security number) that can be used as a unique entity ID, but may also have an email address that needs to be unique across all users in the system.

Reviewer: It may be worth mentioning that this applies not only to ES systems but also to any system without uniqueness constraints in the DB.

Author: I don't think we should go that deep.

> [!NOTE]
> Everything on the side of the `UniqueEmailSubscriber` is happening in the background and independent of the success or failure of the User creation.

### Full Successful Scenario

Now that we have covered the failure scenario, we can have a look at the full picture for the successful scenario:

![Full Successful Flow](flow-full.png?raw=true)

We need to understand that `UniqueEmailSubscriber` and `UserEventsSubscriber` are two autonomous components that
once deployed, run in the background and do their job whenever they receive an event or state change notification.

In the case where a user is successfully created, the `UniqueEmailSubscriber` will still react to the `UniqueEmailEntity` reservation and therefore still schedule a timer to un-reserve it. However, since the user has been created, the `UserEventsSubscriber` will change the `UniqueEmailEntity` state to CONFIRMED and this will trigger another state change notification to the `UniqueEmailSubscriber` which will cancel the timer.

Although simple, this example shows how to implement a **Choreography Saga** in Kalix. We have two entities that 

> influence each other, and we have two actions that listen to events and state changes and guarantee that the whole converges to a consistent state.

Reviewer 2: >This quickstart example demonstrates how to implement a **Choreography Saga** in Kalix. It involves two entities influencing each other, and two actions listening to events and state changes, ensuring the entire application converges to a consistent state.

Move this to a `## Summary` section at the very end of the README.

### Bonus: change an email address

Change the email address of user 001 to `john.doe@acme.com`. Inspect the code to understand how it re-uses the existing saga.

```shell
curl localhost:9000/api/users/001/change-email \
  --header "Content-Type: application/json" \
  -XPUT \
  --data '{ "newEmail": "john.doe@acme.com" }'
```

Check the logs of `UniqueEmailSubscriber` and `UserEventsSubscriber` to see how the saga is progressing.

### check status for email doe@acme.com

```shell
curl localhost:9000/api/emails/doe@acme.com
```

The status of the email will be CONFIRMED or NOT_USED, depending on whether the saga has been completed or not.

## Deploying

To deploy your service, install the `kalix` CLI as documented in
[Setting up a local development environment](https://docs.kalix.io/setting-up/)
and configure a Docker Registry to upload your docker image to.

You will need to update the `dockerImage` property in the `pom.xml` and refer to
[Configuring registries](https://docs.kalix.io/projects/container-registries.html)
for more information on how to make your docker image available to Kalix.

> Finally, you can use the [Kalix Console](https://console.kalix.io)
> to create a project and then deploy your service into the project either by using `mvn deploy kalix:deploy` which
> will conveniently package, publish your docker image, and deploy your service to Kalix, or by first packaging and
> publishing the docker image through `mvn deploy` and then deploying the image
> through the `kalix` CLI.

Reviewer 2: - Do not provide two options for deploying service into project. Eliminate any unnecessary decisions the user has to make.
- Kalix Console cannot be used to create a project.

```suggestion
- Finally, use the `kalix` CLI to generate a project.
- Deploy your service into the project using `mvn deploy kalix:deploy`. This command conveniently packages, publishes your Docker image, and deploys your service to Kalix.
```

Java-SDK [official documentation](https://docs.kalix.io/java/index.html). 

## What is a Choreography Saga?

A **Choreography Saga** is a distributed transaction pattern that helps you manage transactions across multiple services.

In the context of event-driven applications, a **Choreography Saga** is implemented as a sequence of transactions,
each of which publishes an event or state change notification that triggers the next operation in the saga.
If an operation fails because it violates a business rule, then the saga can execute a series of compensating transactions
that undo the changes made by the previous operations.

In Kalix, in addition to events from Event Sourced Entities, you can also subscribe to state changes from Value Entities.
To subscribe to events or state changes, we can use Kalix Actions with the appropriate subscription annotations.

You can create a Choreography Saga to manage transactions across multiple entities in a single service, or across multiple services.
This example implements a choreography that manages transactions across two entities in the same service.

> ## Cross Entity Field Uniqueness

Author: @aludwiko, I reworked this paragraph. You can see the diff [here](https://github.com/lightbend/kalix-jvm-sdk/pull/1884/commits/fc6325ba722257406536bd7e5aec971bdecbd4a8).

Let me know if it's better now.

example, a user may have a unique identifier (e.g. social security number) that can be used as a unique entity ID,
but may also have an email address that needs to be unique across all users in the system.

A **Choreography Saga** can be implemented as a solution to this challenge. In addition to the `UserEntity`, we can 
implement a second entity to act as a barrier. This entity, called the `UniqueEmailEntity`, will be responsible for ensuring that each email address is associated with only one user.  The unique ID of the `UniqueEmailEntity` will be the email address itself. Thus, we can guarantee that there is only one instance of this entity per email.

When a request to create a new `UserEntity` is received, we first check if the email address is in use by first 
trying to reserve it using the barrier, the `UniqueEmailEntity`. If it's not already in use, we proceed to create the `UserEntity`. Once the `UserEntity` has been created, the `UniqueEmailEntity` is marked as CONFIRMED.

If the email address is already in use, the request to create the `UserEntity` will fail.

### Successful Scenario

The sunny day scenario is illustrated in the following diagram:

> ![Successful Flow](flow-successful.png?raw=true)

Author: It depends on the application they use to open the readme file. 

This is a screenshot from VS Code. It also works in IntelliJ.

![CleanShot 2023-11-29 at 15 10 56](https://github.com/lightbend/kalix-jvm-sdk/assets/502982/351ee73d-037c-4a9a-9002-238cbddb062f)

All incoming requests are handled by the `ApplicationController` which is implemented using a Kalix Action. 

1. upon receiving a request to create a new User, the `ApplicationController` will first reserve the email.
2. it will then create the User. 

>    1. the `UserEventsSubscriber` Action listens to the User's event 

Author: Because they don't happen as 3 and 4, but as a consequence of 2. I wanted to express this.

   2. and mark the UniqueEmail entity as soon as it 'sees' that a User has been created.
      
### Failure Scenario

As these are two independent transactions, we also need to consider the failure scenarios. For example, a request to 
reserve the email address may succeed, but the request to create the user may fail. In such a scenario, we end up with 
an email address that is reserved but not associated with a user.

The failure scenario is illustrated in the following diagram:

![Failure Flow](flow-failure.png?raw=true)

1. upon receiving a request to create a new User, the `ApplicationController` will first reserve the email.
   1. in the background, the `UniqueEmailSubscriber` Action listens to state changes from `UniqueEmailEntity`
   2. when it detects that an email was reserved it schedules a timer to un-reserve it after a certain amount of time. 
   3. when the timer fires, the email is un-reserved if the `UniqueEmailEntity` is still in RESERVED state.

> 2. then it tries to create the User, but it fails. As such, the email will never be confirmed, but the timer 

Author: But `I`, `II` and `III` are sub-steps of 1 and dependent on 1. They might happen before 2 or after.

   will unlock it.

> [!NOTE]
> Everything on the side of the `UniqueEmailSubscriber` is happening in the background and independent of the success or failure of the User creation.

### Full Successful Scenario

Now that we have covered the failure scenario, we can have a look at the full picture for the successful scenario:

> ![Full Successful Flow](flow-full.png?raw=true)

Author: Here we are printing the whole flow. In the previous example, we have the explanation of the steps in markdown. And substep (sub-lists) are represented with roman numerals. 

Here I'm not using markdown and then it make it easier to keep arabic numerals. Otherwise, we will have twice `I`.

> ![Full Successful Flow](flow-full.png?raw=true)

Author: The problem here is also that you can not create sub-list in markdown using 2.1, 2.2, etc. If it was possible, it would make our life easy.

> ![Full Successful Flow](flow-full.png?raw=true)

Author: I will revisit the usage of `I`, `II` and `III`. I was also not so happy with it, but couldn't find a better way of expressing it in drawing and in text.

> ![Full Successful Flow](flow-full.png?raw=true)

Author: Revisited.
